bpo-34195: Fix case-sensitive comparison in test_nt_helpers

tjguk opened the pull request:
test_nt_helpers assumed that two versions of a Windows path could be compared case-sensitively. This is not the case, and the difference can be triggered (apparently) by running the test on a path somewhere below a Junction.

This change adds a simple method to compare without considering case and uses it in test_ntpath.TestNtpath.test_nt_helpers

<!-- issue-number: bpo-34195 -->
https://bugs.python.org/issue34195
<!-- /issue-number -->

tjguk committed:
Fix case-sensitive comparison

test_nt_helpers assumed that two versions of a Windows path could be compared case-sensitively. This is not the case, and the difference can be triggered (apparently) by running the test on a path somewhere below a Junction.

tjguk committed:
Merge branch 'master' into issue34195-test-nt-helpers

tjguk committed:
Merge branch 'master' into issue34195-test-nt-helpers

Yhg1s commented on Lib/test/test_ntpath.py:
```diff
@@ -422,15 +422,20 @@ def test_ismount(self):
+    def assertEqualCI(self, s1, s2):
+        """Assert that two strings are equal ignoring case differences
+        """
+        self.assertEqual(s1.lower(), s2.lower())
```
Should this not be str.casefold() instead?

tjguk replied:
Unfortunately, this test checks unicode & bytes. Bytes doesn't support casefold. For the purpose, I'm going to assume that .lower will cover us until someone comes up with a case where it doesn't.

brettcannon commented on Lib/test/test_ntpath.py:
```diff
@@ -422,15 +422,20 @@ def test_ismount(self):
             self.assertTrue(ntpath.ismount(b"\\\\localhost\\c$\\"))
 
+    def assertEqualCI(self, s1, s2):
+        """Assert that two strings are equal ignoring case differences
```
PEP 8 would want this as:
```python
"""Assert that two strings are equal, ignoring case differences."""
```

IOW one line, ending in a period.

tjguk replied:
Done

eryksun commented on Lib/test/test_ntpath.py:
```diff
@@ -422,15 +422,20 @@ def test_ismount(self):
         drive = drive.rstrip(ntpath.sep) + ntpath.sep
-        self.assertEqual(drive, nt._getvolumepathname(sys.executable))
-        self.assertEqual(drive.encode(),
+        self.assertEqualCI(drive, nt._getvolumepathname(sys.executable))
```
This test assumes that `sys.executable` is on the same volume as the path's drive. Say we have a volume mounted at "C:\Mount\Work" and the Python executable is "C:\Mount\Work\CPython\PCbuild\amd64\python_d.exe". In this case `drive` (actually the drive root) is `"C:\\"`, and `nt._getvolumepathname(sys.executable)` is `"C:\\Mount\\Work\\"`. 

The unresolved executable path also might contain multiple junctions or symlinks, such that `_getvolumepathname` returns a path with a different drive. Probably we should be resolving links in `sys.executable` at startup, which IIRC POSIX already does. 

Also, the test assumes the file-system encoding is UTF-8 instead of encoding the path with `os.fsencode`. Python may be configured to use the legacy ANSI encoding. Only recent builds of Windows 10 support setting the ANSI codepage as UTF-8. 

The following should address these concerns:

    executable = nt._getfinalpathname(sys.executable)

    for path in executable, os.fsencode(executable):
        volume_path = nt._getvolumepathname(path)
        path_drive = ntpath.splitdrive(path)[0]
        volume_path_drive = ntpath.splitdrive(volume_path)[0]
        self.assertEqualCI(path_drive, volume_path_drive)

tjguk replied:
Thanks; I've adopted your code

tjguk committed:
Merge branch 'master' into issue34195-test-nt-helpers

tjguk committed:
Make sure docstring is PEP8-compliant
Adopt eryksun's code to account for the fact the sys.executable might be mounted from a different drive

tjguk committed:
Tiny PEP8-related fix
